Extensible context

Simn opened the pull request:
Updated version of https://github.com/Aidan63/haxe/commits/flexible_context/, using an `Array<Any>` as the core data structure.

I'm not adding any of this fancy operator overloading stuff because I don't really get what problem this solves. How exactly the API pans out depends on how all this is supposed to be used and I'm still in the dark about that. I did "port" the scheduler to use this though.

I also don't get why `Element` should be a thing. The only thing this seems to do is associate an object with a key, but I struggle to see in which situation this actually helps. We usually know which kind (Scheduler etc.) of data object we're dealing with, because otherwise we can't do anything with its semantics. The only case this seems to solve is "add this thing to the context but I don't know what it is".

For the keys themselves, I've opted to use `Key.createNew` which has the auto increment built-in. I have a suspicion that once we look into serializing coroutines, we need exact control over key numbers and might not always want to go through the increment mechanic.

This actually makes me realize that this whole `Array<Any>` approach comes with a certain problem: we could have a context with key 0 and key 1000, and a bunch of nothing inbetween. This could be solved with an additional indirection, but I'll have to think about this a bit more. 

Edit: I have no idea why CI fails, that doesn't seem to have to do with anything and doesn't reproduce for me locally...

Aidan63 commented:
> How exactly the API pans out depends on how all this is supposed to be used and I'm still in the dark about that.

I was imagining it looking something like this in the end (taking some of the structured stuff).

```haxe
final threadPool = new FixedThreadPool(4);
final scheduler  = new ThreadPoolScheduler(threadPool);

function main() {
	Coroutine.runScoped(scope -> {

		// First argument is a context to be used with the new scope.
		// The + operator in kotlin allows combining context elements into a context.
		scope.startWith(scheduler + new DebugName('writers'), scope -> {
			for (i in 0...16) {
				// startWith combines the parent context with the context provided to it,
				// so the below coroutines inherit the ThreadPoolScheduler from its parent
				// but the provided DebugName element overwrites the name from the parent context.
				scope.startWith(new DebugName('writer $i'), _ -> {
					foo();
				});
			}
		});

		// do other stuff...

	});
}
```

The context could essentially be considered "coroutine local storage" (might be a better name than context to be honest), each started coroutine gets its own context and will inherit the context elements from its parents unless the element is overridden. The plus operator is just a short hand for creating a context from elements and handles that inheritance vs overridding behaviour, kotlin considers it's contexts' immutable and +ing to them creates new ones (this might be pretty allocation heavy for us).
Hopefully that explains what the end features would be and I've not just re-explained things you already get.

I think there are probably two questions for this.

1. Will this style of flexible context work well in haxe from a user facing api point of view and an implementation one, or have I fell for the trap of following kotlin too strictly again.
2. If not then what haxe friendly mechanism can we come up with to allow this sort of coroutine storage propagation and be open to extension by end users.

Simn commented:
I find this auto-inheriting quite magical... my first thought was that I'd prefer if this was more explicit, but then again it doesn't make much sense to inherit a context without a scheduler, so on some level this makes sense. Interestingly, it doesn't seem very intuitive, but at the same time it's also not very error-prone because it protects users from passing wrong/incomplete context data.

Aidan63 commented:
Yes, I think the auto inheriting stuff is the lesser of two evils. Expecting the users to correctly pass around schedulers, heirarchy coroutines, eventually some cancellation token and whatever else we add seems like a recipe for distaster.
It also opens up adding some kind of `scope` or `scopeWith` coroutine function to create a scope if you don't already have one. e.g.

```haxe
@:coroutine function foo() {
	Coroutine.scopeWith(someScheduler + someOtherElement, scope -> {
		// we now have a scope we can spawn more coroutines in
		// don't need to pass one into foo
	});
}
```

Without some sort of auto inheritance setup the user would need to pass around contexts or their elements manually which seems like a massive pain.
whether we have an abstract with a `+` operator is another thing entirely. I've just been using it in these samples as it's succinct and gets the idea across.

Simn commented:
So I think what we actually want are `clone` and `add` functions which can then be used like this:

```haxe
import haxe.coro.schedulers.EventLoopScheduler;
import haxe.coro.schedulers.Scheduler;
import haxe.coro.context.Key;
import haxe.coro.context.Element;
import haxe.coro.context.Context;
import haxe.coro.Coroutine;

class DebugName extends Element {
	static public var key:Key<DebugName> = Key.createNew("_hx_debugName");
	public var name:String;

	public function new(name:String) {
		super(key.id);
		this.name = name;
	}
}

function startWith(prevCtx:Context, elements:Array<Element>) {
	var newCtx = prevCtx.clone();
	for (element in elements) {
		newCtx.add(element);
	}
	trace(newCtx);
}

function main() {
	var ctx = new Context();
	var scheduler = new EventLoopScheduler(null);
	var debugName = new DebugName("name");
	ctx.add(scheduler);
	ctx.add(debugName);
	startWith(ctx, [new DebugName("new name")]);
}
```

I thought that maybe the allocation of the `elements` Array could be avoided with the `+` thing, but I can't see how that would work without even more magic.

Aidan63 commented:
The allocation is a bit of a bugger, coroutines are suppose to be light weight and you're encourage to start up loads of them. But it seems like the GC pressure could quickly add up.

The only idea on that I've had is that if we say they're immutable that could open up some some easier target specific implementations. e.g. hxcpp could have the context be a special value type which holds GC pointers, so creating a new context is just a new value type and the GC pointers and copied over, no allocations.
But one of the things I like about this third coroutine attempt is that we don't have any target specific stuff and I'm not sure if we want to open that door again...

Simn commented:
I've reworked this to use BalancedTree instead of Array, which should be a good compromise. Lookup will be slightly slower on big contexts, but it's still log(N) and completely avoids the sparse index problem I mention originally. What's also nice about this is that the copying operation only has to allocate the BalancedTree instance, most of the actual data will be shared between instances.

I've also added some `toString` functions which will allow us to print contexts in a way that's useful for debugging. Right now it just ends up as `source/Main.hx:27: Scheduler: [EventLoopScheduler: null], DebugName: [DebugName: new name]`, but we can format this nicer later.

Creating new elements still seems a bit unwiedly:

```haxe
class DebugName extends Element<DebugName> {
	static public var key:Key<DebugName> = Key.createNew("DebugName");
```

Having to explicitly specify both `Element<DebugName>` and `Key<DebugName>` feels very Java... but I don't really know how we could avoid it.

This change apparently also somehow addressed the optimization test failure.

Edit: Also at this point I'm all but convinced that this whole `+` business is pointless and even harmful. The original implementation would clone the entire map for each `+` operation, so you could get a bunch of intermediate allocations that you don't need. By instead passing an `Array<Element>` we only allocate that array once, and then absorb the elements into the context.

Something we could consider is using a rest argument for this instead, because this avoids the array allocation on some targets. However, this would require changing the signature to `startWith(f, ...rest)` which looks weird when `f` is a lambda...

Edit: Perhaps `scope.with(scheduler, new DebugName('writers')).start(f)` could be a thing?

Simn commented:
I'm not sure about the exact class design here, but from a usage perspective this seems quite nice:

```haxe
class AdjustedContext {
	public final scope:Scope;
	public final context:Context;
	public function new(scope:Scope, context:Context) {
		this.scope = scope;
		this.context = context;
	}

	public function start(f:Scope -> Void) {
		f(new Scope(context));
	}
}

class Scope {
	public final context:Context;
	public function new(context:Context) {
		this.context = context;
	}

	public function with(...elements:Element<Any>) {
		var newContext = context.clone();
		for (element in elements) {
			newContext.add(element);
		}
		return new AdjustedContext(this, newContext);
	}
}

function main() {
	var ctx = Context.empty();
	var scheduler = new EventLoopScheduler(null);
	var debugName = new DebugName("name");

	var scope = new Scope(Context.empty());
	scope.with(scheduler, debugName).start(scope -> {
		trace(scope.context.get(DebugName.key).toString()); // [DebugName: name]
		scope.with(new DebugName("new name")).start(scope ->
			trace(scope.context.get(DebugName.key).toString()) // [DebugName: new name]
		);
	});
}
```

This comes out as good looking JavaScript that avoids intermediate allocations due to the `...elements` rest argument:

```js
function Main_main() {
	new Scope(haxe_coro_context_Context.empty()).with(new haxe_coro_schedulers_EventLoopScheduler(null),new DebugName("name")).start(function(scope) {
		console.log("source/Main.hx:57:",haxe_coro_context_Context.get(scope.context,DebugName.key).toString());
		scope.with(new DebugName("new name")).start(function(scope) {
			console.log("source/Main.hx:59:",haxe_coro_context_Context.get(scope.context,DebugName.key).toString());
		});
	});
}
```

On JVM it uses a native array for the rest argument, which is what Java compiles to as well:

```java
    public static void main() {
        Context_Impl_.empty();
        EventLoopScheduler scheduler = new EventLoopScheduler((EventLoop)null);
        DebugName debugName = new DebugName("name");
        Scope scope = new Scope(Context_Impl_.empty());
        scope.with(new Element[]{(Element)scheduler, (Element)debugName}).start((Function)Main_Fields_.Closure_main_0.Main_Fields_$Closure_main_0);
    }
```

Aidan63 commented:
I think this all looks good. It is a shame about the key stuff, I was never a fan of it either but also haven't thought of any way around it.

Simn commented:
Maybe we'll find a better approach down the line, but I'll go ahead and merge this for now so we can use it in #66. I'm quite curious how that API is going to turn out, it feels like the fog is slowly clearing in that regard.
